mkosi/config.py

```diff
@@ -4684,8 +4684,18 @@ def finalize_value(self, setting: ConfigSetting[T]) -> Optional[T]:
         ):
             return v
 
-        if (setting.dest in self.cli or setting.dest in self.config) and isinstance(
-            setting.parse(None, None), (dict, list, set)
+        # If the type is a collection or optional and the setting was set explicitly, don't use the default
+        # value.
+        fieldname = (
+            setting.dest.removeprefix("tools_tree_") if setting.scope == SettingScope.tools else setting.dest
+        )
+        field = Config.fields().get(fieldname)
+        origin = typing.get_origin(field.type) if field else None
+        args = typing.get_args(field.type) if field else []
+        if (
+            (setting.dest in self.cli or setting.dest in self.config)
+            and field
+            and (origin in (dict, list, str) or (origin is typing.Union and type(None) in args))
```

Comment:
Thinking about this again, to only cover `Optional[foo]` this should probably also check `len(args) == 2`, because otherwise something like `None | str | int`, would fall under that, but I think that would be fine in that case as well, so maybe generally.